src/lib/internal.ts

```diff
@@ -35,3 +36,36 @@ export function generateActionObject<T>(
 export function getActive<T>(state: EntityStateModel<T>): T {
   return state.entities[state.active];
 }
+
+/**
+ * Enum that contains all existing Actions for the Entity State adapter.
+ */
+export enum ActionNames {
+  add = 'add',
+  createOrReplace = 'createOrReplace',
+  update = 'update',
+  updateActive = 'updateActive',
+  remove = 'remove',
+  removeActive = 'removeActive',
+  setLoading = 'setLoading',
+  setError = 'setError',
+  setActive = 'setActive',
+  clearActive = 'clearActive',
+  reset = 'reset',
+  goToPage = 'goToPage',
+  setPageSize = 'setPageSize'
+}
+
+/**
+ * An optional annotation to verify that the annotated function has a matching action in ActionNames enum
+ * @see ActionNames
+ */
+export function EntityActionHandler(
+  target: Object, // The prototype of the class
+  propertyKey: string, // The name of the method
+  descriptor: TypedPropertyDescriptor<any>
+) {
+  if (!Object.values(ActionNames).includes(propertyKey)) {
```

Comment:
Will change but shouldn't these be executed once on application startup?

Reply:
Actually I can't use hasOwnProperty, as this checks for keys, but the values are used. I mean so far key and value are identical but the correct way would be to use the values

Reply:
the problem is that your code is less efficient, make the keys the same as the values

Reply:
> Will change but shouldn't these be executed once on application startup?
yes, but checking the value in the array for a long time

Reply:
Is this acceptable?

```typescript
export enum EntityActionType {  // ... 
}

const entityActionTypeValues = Object.values(EntityActionType);

export function EntityActionHandler(
  target: Object, // The prototype of the class
  propertyKey: string, // The name of the method
  descriptor: TypedPropertyDescriptor<any>
) {
  if (!entityActionTypeValues.includes(propertyKey)) {
    throw new NoSuchActionInEnumError(propertyKey);
  }
}
```

Reply:
![image](https://user-images.githubusercontent.com/12021443/50727669-3ad9d400-112f-11e9-862d-40cab1a7a89a.png)

you have the same key name, why don't you just want to do it

```ts
if (!EntityActionType[propertyKey]) {}
```

Reply:
Because the whole goal is to validate things. And all the code is working with the values, not the keys. So it's not logically correct to check with the keys.
That's the whole thought behind this PR. So I don't have to add comments to please not forget to match the keys, values, strings in `generateAction` and method names in `EntityState`.

Reply:
%I threw together a quick performance test:

| Method | Result |
| -- | -- |
| `Object.values` every time | 91% slower |
| Save `Object.values` beforehand | **fastest** |
| `reduce` to object beforehand | 97% slower |
| Use `hasOwnProperty` | **56% slower** |

Reply:
>  validate things
which are object keys

![image](https://user-images.githubusercontent.com/12021443/50728506-e852e480-113b-11e9-997d-97b19bcf29b9.png)

I still don't understand why you need to make an array, if there is a key object anyway

besides, there are faster ways to check for a key
https://jsperf.com/hasownproperty-vs-in-vs-undefined/12

Reply:
@JanMalch Honestly saying, I'm not sure what you are trying to achieve, but this code is slow:

![image](https://user-images.githubusercontent.com/7337691/50728813-762cd080-1138-11e9-98cd-2819da52b096.png)

Let's try to ask V8 to optimize this function:

![image](https://user-images.githubusercontent.com/7337691/50728988-cf95ff00-113a-11e9-8f88-9343ce139f26.png)

V8 even cannot optimize this function :zap:

Reply:
if I have `generateActionObject(ActionNames.setActive ...` the value will be passed into the method, not the key. So you have to use the values. And these values have to match the function names. Not the keys. If I refactor the keys to be capitalized, it won't work anymore, or am I missing something? So you have to use the values everywhere to be consistent.
Of course I know that those are identical so far. But might aswell put hard coded strings again, if you count on them all to be equal.

Reply:
@arturovt I already proposed generating the values array once beforehand and not everytime.

Reply:
I still do not see the point in generating an array if you can do the same keys, but do as you know